File: cellulosesz-api/src/main/java/top/likoslupus/cellulosesz/api/teleport/CellLocation.java
package top.likoslupus.cellulosesz.api.teleport;

import static top.likoslupus.cellulosesz.api.validation.NumericChecks.requireFinite;
import static top.likoslupus.cellulosesz.api.validation.TextChecks.requireNonBlank;

import static java.util.Objects.requireNonNull;

/** Immutable, platform-neutral world position. */
public record CellLocation(
        String world,
        double x,
        double y,
        double z,
        float yaw,
        float pitch
) {

    public CellLocation {
        world = requireNonBlank(requireNonNull(world, "world").trim(), "world");
        requireFinite(x, "x");
        requireFinite(y, "y");
        requireFinite(z, "z");
        requireFinite(yaw, "yaw");
        requireFinite(pitch, "pitch");
    }

    public CellLocation withWorld(String world) {
        return new CellLocation(world, x, y, z, yaw, pitch);
    }

    public CellLocation withPosition(double x, double y, double z) {
        return new CellLocation(world, x, y, z, yaw, pitch);
    }

    public String compact() {
        return "%s %.2f %.2f %.2f".formatted(world, x, y, z);
    }

}

File: cellulosesz-api/src/main/kotlin/top/likoslupus/cellulosesz/api/teleport/CellLocation.kt
package top.likoslupus.cellulosesz.api.teleport

import top.likoslupus.cellulosesz.api.validation.requireFinite
import top.likoslupus.cellulosesz.api.validation.requireNonBlank

/** Immutable, platform-neutral world position. */
@JvmRecord
public data class CellLocation(
    public val world: String,
    public val x: Double,
    public val y: Double,
    public val z: Double,
    public val yaw: Float,
    public val pitch: Float
) {

    init {
        world.requireNonBlank { "world" }
        x.requireFinite { "x" }
        y.requireFinite { "y" }
        z.requireFinite { "z" }
        yaw.requireFinite { "yaw" }
        pitch.requireFinite { "pitch" }
    }

    public fun withWorld(world: String): CellLocation =
        CellLocation(world, x, y, z, yaw, pitch)

    public fun withPosition(x: Double, y: Double, z: Double): CellLocation =
        CellLocation(world, x, y, z, yaw, pitch)

    public fun compact(): String =
        "%s %.2f %.2f %.2f".format(world, x, y, z)

    public fun isFinite(): Boolean {
        return world.isBlank()
                && java.lang.Double.isFinite(x)
                && java.lang.Double.isFinite(y)
                && java.lang.Double.isFinite(z)
                && java.lang.Float.isFinite(yaw)
                && java.lang.Float.isFinite(pitch)
    }

}
